File: pat1038.cpp
#include<iostream>
#include<string>
#include<algorithm>
using namespace std;

bool comp(string s1, string s2) {
	return s1 + s2 < s2 + s1;
}

int main() {

	int n;
	cin >> n;
	string segments[10010];

	for (int i = 0; i < n; i++) {
		cin >> segments[i];
	}

	sort(segments, segments + n, comp);

	string result = "";
	// erase 0 in front of the first number
	if (segments[0][0] == '0') {
		result = segments[0].substr(1);
	}

	for (int i = 1; i < n; i++) {
		result += segments[i];
	}

	if (result.length() == 0) {
		cout << 0;
	}
	else {
		cout << result;
	}

}

File: pat1038.java
//      attention: test case 6 Time Limit Exceeded

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.Comparator;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        String[] line = reader.readLine().split(" ");
        Arrays.sort(line, 1, line.length, new Comparator<String>() {
            @Override
            public int compare(String o1, String o2) {
                return (o1 + o2).compareTo(o2 + o1);
            }
        });


        String result = "";

        for (int i = 1; i < line.length; i++) {
            result += line[i];
        }

        // erase 0 in front of the number
        result = result.replaceFirst("^0*", "");

        if (result.length() == 0) {
            System.out.println(0);
        } else {
            System.out.println(result);
        }
    }
}
